README
This is FLUX, the Field Line Universal Relaxer, v. 2.2

For full documentation visit the FLUX wiki:
     http://flux.boulder.swri.edu/


FLUX is a prototype MHD solver that is currently able to find
nonlinear force-free fields with an exactly specified topology in a
quasi-static, infinitely conductive, low-beta medium.  Unlike
conventional extrapolation codes, FLUX accepts as input a collection
of field line shapes that do not necessarily have physical meaning but
that do have the correct topology that you want to study.

The FLUX engine relaxes the position of those field lines until they
approximate the physical force-free state with the same topology that you
specified.

FLUX consists of two parts: a relaxation engine, written in C, and a
control interface in Perl.  The control interface is intended to be
used with Perl Data Language (PDL).  If you do not have PDL on your
system, you can get it for free from http://pdl.perl.org.  PDL
compiles "out of the box" on most UNIX systems, and a "native" 
version exists for Microsoft Windows.

The authors gratefully acknowledge support from NASA's Living With A
Star and Solar/Heliospheric Physics programs, and from the Southwest
Research Institute.

All source files are copyright (C) 2004-2008 by the authors
unless otherwise marked.  You may modify and/or make copies of this
software under the terms of the Gnu Public License, version 2 (GPL
v2).  You should have received a copy of the GPL v2 with this
software; it is present in the file "COPYING".  If you do not have a
copy you may obtain one online from "http://www.gnu.org" or by writing
to the Free Software Foundation, 59 Temple Place, Suite 330, Boston,
MA 0211-1307 USA.

THIS SOFTWARE COMES WITH NO WARRANTY OF ANY KIND.

Authors:
	Craig DeForest	 	deforest@boulder.swri.edu
	Charles Kankelborg 	kankel@solar.physics.montana.edu
	Laurel Rachmeler	laurel@boulder.swri.edu
	Alisdair Davey		adavey@spd.org

Contributors (gratefully acknowledged):
	     Nathan Schwadron
	     Teddy Walls
	     Christopher Plumberg

--------

This is the FLUX top-level directory.

Subdirectories include:
        lib -- the central libflux.a core simulation library
		    (this gets installed into /usr/local/lib/)

	doc -- some docs (deprecated; see http://flux.boulder.swri.edu 
		for more up-to-date information.)

	include -- 'c' header files 
	           (these get installed into /usr/local/include/flux/)
	
        pdl -- PDL glue code and perl modules

	pdl/PDL -- PDL autoload subroutines

RELEASE NOTES FOR v2.2 (22-Nov-2008):
  - Mass is actively tracked in the data structures
  - Binary save files are supported; all files are auto-gzipped if they have a '.gz' extension
  - First-stage parallelization appears functional -- the code can spawn to make use of multiple CPU cores on the local machine.
  - Interpolation of all scientific values onto arbitrary locations is now possible.
  - New reconnection criteria make reconnection "in the wrong direction" much less common.
  - Multiple bug fixes in the perl interface code
  - Data structure labels can now be edited via the perl hash interface.
  - There is better support for motion and interaction of flux concentrations.

RELEASE NOTES FOR v2.1 (3-Apr-2008):
	
  - Many, many bug fixes
  - Photospheres and other refs handled correctly by hash interface
  - Reworked Perl hash interface to eliminate stale pointers altogether
  - Added Hilbert ditherer
  - Better reconnection support
  - Better auto-open support

RELEASE NOTES FOR v2.0 (1-Nov-2007):

  - Now includes reconnection via the "reconnect" Flux::World method
  - Updated and validated force laws
  - Full plasmoid support
  - Full open-field boundary support, including omega and U loop
	interactions with the boundary
  - Includes a Floyd-Steinberg ditherer for footpoint placement
  - Includes techniques ("ductions") for moving footpoints in 
	time-dependent simulations
  - Better control over relaxation parameters
  - Emergence/cancellation support
  - Better interaction with Perl:
	- Support for all fields via the Perl hash interface
	- No memory leaks (I think)
	- No stale pointers (I think)



Changes
12-Aug-2010[cd
	- fix the label hasher for new vertices.  This fixes the not-reading-in-neighbors-from-save-files bug.
9-Nov-2009[cd]
	- allow "(null)" as fluxon end-condition name
29-Oct-2009[cd]
	- fully parallelized the two-stage stepper.  This is important because the final portion 
(fluxon_relax_step) potentially runs slowly since it involves walking out into the neighbor tree.
9-July-2009[lr]
	-fiddled with the reconnection conditions. These are not yet done, use at your own risk. 
15-Apr-2009[lr]
	-added make_world_gh.pdl and gold_hoyle_line.pdl to pdl/PDL.
6-Mar-2009[lr]
	-changed the global update force. Old way is still there and still works with the '1' global number. Simple relaxer now uses -1 on the first go that does a fast global neighbor search instead.
10-Feb-2009 [cd]
	    * new_vertex generates random vertex IDs to support very large arrays -- 
	      sequential ID generation is pessimal for binary-tree insertion.
	      new_vertex also only balances the tree once every thousand vertices.  
	      This should probably be matched with an explicit tree-balancing operation
	      sprinkled around after most new-vertex operations, but is not (yet).
20-Jan-2009 [lr]
	    Added several routines in model.c and geometry.c to handle calculating the hull of a begin/end vertex on the photosphere. All the routines are all the bottom, and are not part of the regular relaxation step. They are only called through PDL ($v->photohull). Also added to vertex.xs to impliment this.
	    incremented the FLUX_CORE_VERSION to 17 in fluxperl.h
20-Jan-2008 [cd]
	    Minor fixes to try avoiding reconnection problems with neighbors
6-Dec-2008 [cd]
	   Added FL_ISDUMMY; converted a bunch of ad-hockery to FL_ISDUMMY calls.

----------
RELEASED VERSION 2.2
----------
20-Nov [cd]
       checked-in local-maximum reconnection criteria
19-Nov [cd]
       Fixed problem with rc_funcs setting -- increment refct on return val.
18-Nov [cd]
       Spliced mass handling into the  main data structures; added f_v_ps field for 
       pseudoforces.  No I/O or physics handling yet - the quantities are just tracked.
11-Nov-2008 [cd]
    - wonky-vertex problem solved: duplicate neighbors were not being deleted in the parallel
      case, so operations like opening fluxons (which creates duplicate neighbor candidates
      in the newly-opened fluxon) were causing duplicate neighbors.  The hull algorithm works
      correctly -- it was the merge-neighbors operation in io.c (for parallel operation) that
      was not working properly.
6-Nov-2008 [cd]
    - Debugging of wonky-vertex problem in place (some vertices develop HUNDREDS of neighbors!)
    - interpolation of values works now using the interpolate_value method in Flux::World.
3-Nov-2008 [cd]
    - binary file save architecture in place (hopefully runs faster than ASCII?)
    - Use parallelization on initial update_neighbors.
    - Per-vertex globalization of neighbor finding should avoid the need for global neighbor finding to initialize -- code detects which vertices require global finding.
    - Added vertex neighbor dumblist size/n to vertex hash interface
23-Oct-2008 [cd]
    Updated fix_curvature issues (r_ncl now transmitted via io.c)
22-Oct-2008 [cd]
    First-stage parallelization appears complete.  Control number of CPUs with the {concurrency} 
    field of a Flux::World object.  Setting it to 0 uses no concurrency; setting it to 1 uses
    the parallel code (with the piping penalty that implies) but only allows it to use one CPU.
    Setting it to a large number (on an appropriately equipped workstation) appears to accelerate the code 
    by a large factor.
    
    Other fixes:
    	  -io.c: a protocol for binary file transfer (only the parts needed for parallelization are 
	  	 finished)
    	  -overhauled the force/curvature monitoring code: now there is an encapsulated accumulator in 
	  	      model.c
          -The maximum and mean diversion angle are now reported in degrees, not cos(theta).

21-Oct-2008 [cd]
	    First cut at first-stage parallelization...
20-Oct-2008 [cd]
	Added several new utilities toward parallization:
	      io.c: a binary-save-file infrastructure that uses machine order and encoding
	      data.h, data.c: added a "concurrency" field to the WORLD structure
	      perl side: added interface code for the concurrency field.
7-Oct-2008 [cd]
	Several utility additions to manipulate fluxon/concentration connections on the perl side...
26-Sep-2008 [cd]
	Added vertex label-change interface to perl side...
30-July-2008 [cd]
        added world_check to model.h; perl interface in World.xs.  Updated docs
	for render() in World.pm.  read_world now automatically calls world_check 
	to fix up questionable incoming files.
29-July-2008 [cd]
	io.c - fixed bug in reading that was scrozzling flux values for fluxons
	(and hence screwing up magnetic field values).
17-July-2008 [cd]
	World.xs, Core.xs, fluxperl.h, geometry.h, geometry.c -
	  installed closest_vertex and closest_simplex perl methods and
	  their associated c counterparts find_vertex_by_location and 
	  find_simplex_by_location -- these will hopefully come in handly
	  later for interpolating values onto a grid!
17-July-2008 [lr]
  - changed f_vertex5 in physics.c to inclue all 3 forces: proximity attraction, vertex repulsion, and curvature attraction.
  - added a new routine to put in straight fluxons that end on an open sphere: vertical_fluxons. 
15-July-2008 [cd]
  (Missed a few changes?)
	World.xs - relative helicity finder is in place
		 - vertex finder-by-location is present but not yet working.
        io.c - fixed a problem with autogenerated vertices colliding with 
	       special vertices.
18-Apr-2008 [lr]
  - Second photosphere isnow working. Added a 2nd image fluxon and 
    2 more image charges to handle the image correctly.
  - Fixed a bug that had been alluding us for a while. When a new 
    vertex is created because the old one wanted to step through the 

include/data.h
/* Includes for FLUX -- field line and tie point structures 
 *
 * This file is part of FLUX, the Field Line Universal relaXer.
 * Copyright (c) Craig DeForest, 2004-2008
 * 
 * You may modify and/or distribute this software under the terms of
 * the Gnu Public License, version 2.  You should have received a copy
 * of the license with this software, in the file COPYING to be found
 * in the top level directory of the distribution.  You may obtain
 * additional copies of the license via http://www.gnu.org or by
 * writing to the Free Software Foundation, 59 Temple Place - Suite
 * 330, Boston, MA 02111-1307 USA.
 *
 * The software comes with NO WARRANTY.
 * 
 * You may direct questions, kudos, gripes, and/or patches to the
 * author, Craig DeForest, at "deforest@boulder.swri.edu".
 * 
 * This file is part of FLUX 2.2 (22-Nov-2008).
 */

#ifndef FLEM_DATA
#define FLEM_DATA 1

#define NUM double             /* Type of calculation.  Double is (in these  */
#define NUMDOUBLE 1            /* days of 64-bit processors) actually faster*/
                               /* than float, but uses twice as much memory!*/
/* NUM must be one of (float|double|longdouble).  Set NUMDOUBLE to (0|1|2),
 * and NUMCHAR to "", "l", or "L". [NUMCHAR modifies NUM parsing by printf --
 * the 'l' makes the parse a double.]
 */
#define NUMCHAR "l"

#include <stdlib.h>

/* MD5 hashing is now used for new VERTEX ID's CED 10-Feb-2009 */
#include <time.h>
#include <unistd.h>

/**********************************************************************
 * Data structures
 * 
 * Note that VERTEXs and FLUXONs both contain different sorts
 * of tree linking.  The trees are dealt with by a bunch of semi-generic
 * tree methods.  The tree methods keep track of a running sum of
 * both number of links and also magnetic flux.  The semi-genericity
 * comes from the variable offset from the struct address of the link
 * table used for the tree.  One of the running sums comes from the
 * magnetic flux (or any other NUM), which must be the very first item
 * in the struct.
 *
 * The VERTEXs and FLUXONs contain their own tree links to facilitate
 * moving between different data trees.  There is also a generic tree
 * construct which is used for passing around bunches of things.
 *
 */

typedef NUM POINT2D[2];
typedef NUM POINT3D[3];

typedef struct LINKS {
  void *left;
  void *right;
  void *up;
  long n;
  NUM sum;
  char balanced;
} LINKS;
  

typedef struct TREE {
  void *node;  /* node pointer serves as label! */
  LINKS links;
} TREE;
  

#ifndef NULL
#define NULL ((void *)0)
#endif

typedef struct DUMBLIST {
  void **stuff;
  unsigned int n;
  unsigned int size;
} DUMBLIST;

typedef struct VERTEX {
  struct FLUXON *line;                   /* Which line are we on? */
  struct VERTEX *prev,*next;  

     /* Vertex location in physical space */
  POINT3D x;           

     /* neighbors & nearby - infrastructure for geometrical work */
  DUMBLIST neighbors;
  DUMBLIST nearby;

     /* Scratch space used when this vertex is considered as a neighbor of
      * someone else; part of force calculation. */
  POINT3D scr; 
  long passno;  /* see geometry.h, geometry.c (anti-repeat gate) */
  NUM r,a;      

     /* Magnetic field info is calculated by b_vec in physics.c.   */
     /* Remember that it is segment-centered, not vertex-centered. */
  POINT3D b_vec;                /* Stores the local B vector */
  NUM     b_mag;                /* Stores the magnitude of the B field */
  NUM energy;                            /* calculated B energy for segment */

     /* Force accumulators are filled by the force functions in physics.c. */
     /* As of v3, the vertex pseudoforce can be tracked indepenently of the */
     /* plasma force that is localized at the vertex.  For the purpose of relaxation */
     /* step sizing, the f_v_ps magnitude is added to f_v_tot, not to a separate variable. */
  POINT3D f_v;                  /* Stores force on the vertex */
  POINT3D f_v_ps;               /* vertex pseudoforce - used for vertex distribution */
  POINT3D f_s;                  /* Stores force on the following segment */
  POINT3D f_t;                  /* Stores total force */
  NUM f_s_tot;                  /* sum-of-magnitudes force on the segment (for help in auto-step-size)*/
  NUM f_v_tot;                  /* sum-of-magnitudes force on the vertex  (for help in auto-step-size)*/
  NUM f_n_tot;                  /* norm of the total force */


  NUM r_v, r_s;                 /* Stores relevant lengthscale for vertex & segment forces */
  NUM r_cl;                     /* Stores closest-approach radius of any vertex */
  NUM r_ncl;                    /* Stores closest-approach radius of neighbors */

  long label;                  
  LINKS world_links;      /* tree access to all the vertices in the world */
  POINT3D plan_step;       /* planned step */


  /* Single-fluid data -- these are valid halfway down the following segment. */
  NUM rho;         // Density 
  NUM T;           // Internal energy
  POINT3D p;       // Momentum

} VERTEX;

#define V_ISDUMMY(v) ( ((v)->label <= -1 && (v)->label >=-4) || (!(v)->line) || !((v)->line->label) )

typedef struct FLUXON {
  NUM flux;                  /* Flux in Maxwells */
  long label;                /* Unique ID assigned at creation */
  

  struct VERTEX *start, *end; /* Endpoints of the field line (could be the

		                   same point for loops!!) */
  long v_ct;         /* Index counter -- should be the number
			 of VERTEXes in the fluxon.  */

  
  LINKS all_links;      /* structure containing all fluxons */
  LINKS start_links;    /* structure of fluxons sharing this one's start */
  LINKS end_links;      /* structure of fluxons sharing this one's end */
  struct FLUX_CONCENTRATION *fc0, *fc1;
  char plasmoid;

} FLUXON;

#define FC_OB_ID  -1
#define FC_OE_ID  -2
#define FC_PB_ID  -3
#define FC_PE_ID  -4
#define FC_IM10_ID -8
#define FC_IM11_ID -9
#define FC_IM20_ID -10
#define FC_IM21_ID -11
#define FC_ISDUMMY(fc) ( (fc)->label < 0 && (fc)->label >= -11 )

#define FL_IM1_ID -9
#define FL_IM2_ID -10
#define FL_ISDUMMY(fl) ( (fl)->label == -9 || (fl)->label == -10 )

typedef struct FLUX_CONCENTRATION {
  struct WORLD *world;
  NUM flux;                  /* Total magnetic flux in Maxwells */
  long label;                /* Unique ID assigned at creation */
  struct FLUXON *lines;      /* Tree of lines coming out */
  struct LINKS links;        /* Tree for other concentrations */
  NUM x[3];                  /* Location of the concentration */
  NUM locale_radius;         /* Radius of the concentration's neighborhood */
  
  long passno;               /* Multiple-pass detection for hull calculation */
  void (*bound)(VERTEX *v);  /* Boundary condition routine */
  void (*pbound)(VERTEX *v); /* Plasma boundary condition routine */
  
} FLUX_CONCENTRATION;

typedef struct PLANE {
  NUM origin[3];
  NUM normal[3];
} PLANE;

#define PHOT_NONE 0
#define PHOT_PLANE 1
#define PHOT_SPHERE 2
#define PHOT_CYL 3
extern char *BOUNDARY_NAMES[4];  /* in data.c */
	

typedef struct PHOTOSPHERE {
  PLANE *plane;
  int   type;
} PHOTOSPHERE;

#define WORLD_STATE_NEW     0
#define WORLD_STATE_LOADING 1
#define WORLD_STATE_LOADED  2
#define WORLD_STATE_WORKING 3
#define WORLD_STATE_RELAXED 4
#define WORLD_STATE_READY   5

#define MAXNUMCOEFFS 1024 /* Number of coefficients to create */

#define N_FORCE_FUNCS 30    /* Number of functions allowed in the force list */
#define N_RECON_FUNCS 10    /* Number of reconnection condition functions in the recon list */
#define N_RECON_PARAMS 8    /* Number of world-global parameters to be stored for each reconnection function */

#define N_M_PARAMS 5        /* Number of world-global parameters for a generic mass scaling function */

#define N_NEIGHBORS_PREALLOC 10 /* Number of neighbors to pre-allocate for each vertex.  Saves time on file read. */

typedef VERTEX *((RC_FUNC)(VERTEX *v, NUM *params));

typedef struct WORLD {
  /* Globally memorable fields go here */
  long frame_number;        /* Identifies currently-being-worked-on time step */
  long state;               /* State of the world  */
  long refct;               /* Used for perl interface (deallocate world when this reaches zero) */
  FLUX_CONCENTRATION *concentrations;
  FLUXON *lines;
  VERTEX *vertices;

  PHOTOSPHERE photosphere;
  PHOTOSPHERE photosphere2;

  FLUX_CONCENTRATION *fc_im10,*fc_im11,*fc_im20,*fc_im21;
  FLUXON *fl_im,*fl_im2;
  VERTEX *image, *image2,*image3 ,*image4;

  NUM locale_radius;        /* Default radius for concentrations' neighborhoods */

  long auto_open;             // Flag indicating whether a full-auto open boundary is being maintained
  
  FLUX_CONCENTRATION *fc_ob; /* bogus flux conc. to store open fluxons     */
  FLUX_CONCENTRATION *fc_oe; /* bogus flux conc. to store open fluxons     */
  FLUX_CONCENTRATION *fc_pb; /* bogus flux conc. to store plasmoid fluxons */
  FLUX_CONCENTRATION *fc_pe; /* bogus flux conc. to store plasmoid fluxons */

  void (*default_bound)(VERTEX *v);  /* Boundary condition routine */

  long verbosity;           /* Verbose flag turns on/off debugging lines */
  
  void ((*(f_funcs[N_FORCE_FUNCS]))());         /* force functions (incl. B field etc.) */
  void ((*(m_funcs[N_FORCE_FUNCS]))());         /* mass calculation functions */
  RC_FUNC *rc_funcs[N_RECON_FUNCS];             /* reconnection threshold detectors */
  NUM rc_params[N_RECON_FUNCS][N_RECON_PARAMS]; /* reconnection threshold parameters */

  NUM m_params[N_M_PARAMS];
  
  struct { 
    NUM b_power; 
    NUM d_power;
    NUM s_power;
    NUM ds_power;
  } step_scale;

  int passno;
  int handle_skew; // flag used to tweak Voronoi cell wall projection; see geometry.c

  // ARD added code to keep largest angle between fluxels and mean angle
  // between fluxels for WORLD 

  NUM max_angle;
  NUM mean_angle;
  // ARD - world should keep step and dt 
  NUM dtau;
  long rel_step;
  // ARD - array of coefficients 
  NUM coeffs[MAXNUMCOEFFS];
  long n_coeffs;
  long maxn_coeffs;

  // CED - how many separate processes should be run at once when calculating forces?
  long concurrency; 

  // Force and angle accumulators for global statistics;
  NUM f_min;
  NUM f_max;
  NUM fr_min;
  NUM fr_max;
  NUM ca_min;
  NUM ca_max;
  NUM ca_acc;
  NUM ca_ct;

  // Switch use of mass
  int use_fluid;

  // temporary dumblist used for I/O (vertex number translation)
  DUMBLIST *tmplist;


} WORLD;

const char *world_state_name(WORLD *a);

/**********************************************************************
 * Barf opcodes
 */

#define BARF_MALLOC 1

/**********************************************************************
 * Routines in data.c
 */

void barf(int barf_on_op, char *where);

long new_label(long request);
long new_vertex_label(long request);
long hash_vertex_label(long request, WORLD *w);

/* Offsets defined here... */
static const FLUXON f__samp;
static const fl_lab_of  = (long)&(f__samp.label) - (long)&f__samp;
static const fl_all_ln_of = (long)&(f__samp.all_links) - (long)&f__samp;
static const fl_start_ln_of = (long)&(f__samp.start_links) - (long)&f__samp;
static const fl_end_ln_of = (long)&(f__samp.end_links) - (long)&f__samp;

static const FLUX_CONCENTRATION fc__samp;
static const fc_lab_of = (long)&(fc__samp.label) - (long)&fc__samp;
static const fc_ln_of = (long)&(fc__samp.links) - (long)&fc__samp;

static const TREE tree__samp;
static const tree_lab_of = (long)&(tree__samp.node) - (long)&tree__samp;
static const tree_ln_of  = (long)&(tree__samp.links) - (long)&tree__samp;

static const VERTEX v__samp;
static const v_neighbor_of = (long)&(v__samp.neighbors) - (long)&(v__samp);
static const v_nearby_of = (long)&(v__samp.nearby) - (long)&(v__samp);
static const v_ln_of = (long)&(v__samp.world_links) - (long)&(v__samp);
static const v_lab_of = (long)&(v__samp.label) - (long)&(v__samp);
 FLUXON *new_fluxon(NUM flux,  
			 FLUX_CONCENTRATION *c0,
			 FLUX_CONCENTRATION *c1,
			 long label,
			 char plasmoid
			 );

 VERTEX *new_vertex(long label, NUM x, NUM y, NUM z,
			      FLUXON *fluxon);

 VERTEX *spawn_new_vertex(NUM x, NUM y, NUM z, VERTEX *parent);

       int add_vertex_pos     (FLUXON *fl, long pos,  VERTEX *v);
 int add_vertex_after   (FLUXON *fl, VERTEX *nbr,  VERTEX *v);



 FLUX_CONCENTRATION *new_flux_concentration(WORLD *, NUM x, NUM y, NUM z,
						  NUM flux,
						  long label
						  );
void check_special_concentration(FLUX_CONCENTRATION *f);

WORLD *new_world();

void free_world( WORLD *w );
void delete_flux_concentration(FLUX_CONCENTRATION *fc);
void delete_fluxon(FLUXON *f);
void delete_vertex(VERTEX *v);
void unlink_vertex(VERTEX *v);
void vertex_clear_neighbors(VERTEX *v);
void vertex_add_neighbor(VERTEX *v, VERTEX *neighbor);
int vertex_renumber(VERTEX *v, long newlab);
int fluxon_renumber(FLUXON *f, long newlab);
int concentration_renumber(FLUX_CONCENTRATION *fc, long newlab);

#define MAX_TREE_DEPTH 1000000
void clear_links(LINKS *links);
 void *tree_top(void *tree, int link_offset);
 void *tree_find(void *tree, long label, int label_offset, int link_offset);
 void *tree_insert(void *root, void *item, int label_offset, int link_offset);
 void *tree_binsert(void *root, void *item, int label_offset, int link_offset);
 void *tree_unlink(void *data, int label_offset, int link_offset);
 void *tree_bunlink(void *data, int label_offset, int link_offset);
 void *tree_balance(void *tree, int label_offset, int link_offset);
 char tree_balance_check(void *tree, int link_offset);

 long tree_walk(void *tree, int label_ofset, int link_offset, long ((*func)()));
long tree_walker(void *tree, int label_ofset, int link_offset, long ((*func)()),int depth);

long stw_helper(void *node, int lab, int link, int depth);
long safe_tree_walker(void *tree, int label_offset, int link_offset, long((*func)()), int depth);


DUMBLIST *new_dumblist();
void free_dumblist(DUMBLIST *dl);
void dumblist_init(DUMBLIST *foo);
void dumblist_clean(DUMBLIST *foo);
void dumblist_add(DUMBLIST *dl, void *a);
void dumblist_quickadd(DUMBLIST *dl, void *a);
void dumblist_delete(DUMBLIST *dl, void *a);
void dumblist_rm(DUMBLIST *dl, int i);
void dumblist_sort(DUMBLIST *dl, int((*compare)(void *a, void *b)));
void dumblist_snarf(DUMBLIST *dl, DUMBLIST *source);
void dumblist_grow(DUMBLIST *dl, int size);
void dumblist_clear(DUMBLIST *dl);
void dumblist_crunch(DUMBLIST *dl);

char fl_eq(NUM a, NUM b);


/* defines for debugging malloc wrapper in data.c */
#define MALLOC_WORLD 1
#define MALLOC_FLUXON 2
#define MALLOC_VERTEX 3
#define MALLOC_FC 4
#define MALLOC_DL 5
#define MALLOC_DL_L 6
#define MALLOC_VL 7
#define MALLOC_DLS_ARENA 8
#define MALLOC_PLANE 9
#define MALLOC_MISC 10
#define MALLOC_MAXTYPENO 10



char *malloc_types[MALLOC_MAXTYPENO+1];

#define valid_malloc_type(x) ((x)>0 && (x)<=MALLOC_MAXTYPENO)


/****************************************
 * Memory handling options
 * We have the ability to use either vanilla malloc,
 * a debugging/fencing malloc, or a perl-SV-based malloc.
 * If libflux is linked to the PDL front end, then 
 * you should use the perl malloc.
 */

#ifdef USE_DEBUGGING_MALLOC

#define localmalloc(x,y) flux_malloc(x,y)
#define localfree(x) flux_free(x)

void flux_free(void *p);
char *flux_malloc(long size, int what_for);
void flux_memcheck();

#else
#ifdef USE_PERL_MALLOC

#define localmalloc(x,y) flux_perl_malloc(x)
#define localfree(x) flux_perl_free(x)

 char *flux_perl_malloc(long size);
 void flux_perl_free(void *where);

#include "EXTERN.h"
#include "perl.h"
#include "perlapi.h"

#else
#ifdef USE_PADDED_MALLOC

#define localmalloc(x,y) flux_padded_malloc(x)
#define localfree(x) flux_padded_free(x)

 char *flux_padded_malloc(long size);
 void flux_padded_free(void *ptr);
#else

/*** Default case: just use normal malloc ***/
 #define localmalloc(x,y) malloc(x)
 #define localfree(x) free(x)

#endif /**** use_padded_malloc test ***/
#endif /**** use_perl_malloc test ****/
#endif /**** malloc switch ****/
#endif /* overall file include */


/******************************
 *
 * version-tracking: informational strings about each file in use
 */
extern char *code_info_model;
extern char *code_info_physics;
extern char *code_info_io;
extern char *code_info_geometry;
extern char *code_info_data;
